### DISCUSSION TITLE
Minima and maxima for dimensionality reduction

### DISCUSSION THREAD

**Discussion by boris-kz** [boris-kz/CogAlg#40] [Ideas]
I was thinking of using minima and maxima for blob dimensionality reduction (right now we are using blob median for that). 
It's a common practice and obviously useful in many cases, but I haven't seen any conceptual justification. 
Now it seems me that's because extrema are generally more stable than other points, and stability is our core criterion.
I understand why this so in topographical maps: stability is inverse to erosion, which is proportional to slope.
Extrema have 0 slope, thus the slowest erosion.
But I don't know how to generalize this beyond topography?

This is regarding "unsupervised" value, leaving aside utilitarian value: these extrema are usually transportation obstacles, etc.

**Comment by kwcckw** [comment]
> I was thinking of using minima and maxima for blob dimensionality reduction

The dimension here is referring to the actual pixel size or parameter size?

> Now it seems me that's because extrema are generally more stable than other points, and stability is our core criterion.
> I understand why this so in topographical maps: stability is inverse to erosion, which is proportional to slope.
> Extrema have 0 slope, thus the slowest erosion.
> But I don't know how to generalize this beyond topography?

For the generalization, are you look into converting this idea into some function so that we can use it for the dimension reduction purpose?

**Reply by boris-kz** [reply]
2D blob -> 1D ridge of maxima -> 0D point maximum.
I am looking for a justification in terms of maximizing overall match.
Sorry, this is probably too abstract.

**Reply by kwcckw** [reply]
i see, yea it's quite abstract , but how would the dimension reduction correlated to maximize the overall match?

**Reply by boris-kz** [reply]
Dimensionality reduction is a form of compression: you replace multiple parameters with one that is still a good representation of the object. And my match is a measure of compression.

Ridge is not only compressed relative to blob, it's also more stable under various conditions. Stable means matching across temporal or spatial distance, despite the fact that change is generally proportional to these distances.

The question is, is the ridge of maxima theoretically more stable than median line? Median is directly computed from the whole blob, which is defined by internal match. Maxima are not directly related to other points in the blob, so they need some indirect justification.

**Reply by kwcckw** [reply]
Thank you for the explanations. 

So in this case, you are speculating whether a ridge of maxima is more stable than the median line right? Where the median line is defined by the blob internal match. 

I think we may need to find some related papers to get more related information.

From my initial search, i can only find this but is this relevant enough?

https://www.researchgate.net/publication/279468264_On_the_Behaviour_in_Scale-Space_of_Local_Extrema_and_Blobs

**Reply by boris-kz** [reply]
Thanks Chee, I will take a look!

**Reply by kwcckw** [reply]
> 
> 
> Thanks Chee, I will take a look!

Welcome, but it's actually an old paper, and i don't see much new paper related to this now. Anyway, i will search further some time later again.

**Reply by kwcckw** [reply]
found a few more related materials, might or might not related, need go through them first:

Scale Selection Properties of Generalized Scale-Space InterestPoint Detectors
https://core.ac.uk/download/pdf/204897319.pdf

Scale-space blob
https://books.google.com.my/books?id=KzfoBwAAQBAJ&pg=PA193&lpg=PA193&dq=extrema+point+and+blobs&source=bl&ots=dgWDTgzGf4&sig=ACfU3U36ZOgCeEVT3qEmtPJxypjYrU01tg&hl=en&sa=X&ved=2ahUKEwjW_9yW3rHuAhV56nMBHTeNBoo4FBDoATAEegQIBRAC#v=onepage&q=extrema%20point%20and%20blobs&f=false

The Topological Structure of Scale-Space Images
https://www.researchgate.net/publication/2534355_The_Topological_Structure_of_Scale-Space_Images

**Reply by boris-kz** [reply]
> Thank you for the explanations.
> 
> So in this case, you are speculating whether a ridge of maxima is more stable than the median line right? Where the median line is defined by the blob internal match.
> 
> I think we may need to find some related papers to get more related information.
> 
> From my initial search, i can only find this but is this relevant enough?
> 
> https://www.researchgate.net/publication/279468264_On_the_Behaviour_in_Scale-Space_of_Local_Extrema_and_Blobs

So, this looks like observational analysis, not deductive one. It maybe suggestive but doesn't answer the "why" question.

**Reply by kwcckw** [reply]
Right, it is more like stating the behaviour of those critical points under different conditions, and they only answer 'why' in the last few proposition by clarifying it with some proof.

However, some features are using extrema as point of interest or feature, i think because it's more stable, such as:

SIFT

![image](https://user-images.githubusercontent.com/52521979/105620063-eb5cc280-5e33-11eb-9cf8-a13e90502e11.png)

Link:
https://www.researchgate.net/profile/Ebrahim_Karami/publication/318194587_Image_Matching_Using_SIFT_SURF_BRIEF_and_ORB_Performance_Comparison_for_Distorted_Images/links/5965089b0f7e9b2a367cd913/Image-Matching-Using-SIFT-SURF-BRIEF-and-ORB-Performance-Comparison-for-Distorted-Images.pdf 



Canny edges:

![image](https://user-images.githubusercontent.com/52521979/105620098-3d9de380-5e34-11eb-85a0-c3246089df47.png)

Link:
https://www.researchgate.net/publication/233834337_Scale-space_color_blob_and_ridge_detection

**Reply by boris-kz** [reply]
- I don't want to use Gaussians because it's a form of resolution reduction. I think resolution (a type of my filters) should only be reduced after evaluation per pattern, which means after a full cross-comp and clustering cycle. So, scale-space here is compositional hierarchy of patterns, not a spectrum of Gaussian kernels.

- Key points in SIFT are just another name for extrema? They may be more stable than blob averages / medians because their gradient (of the parameter over which the extrema is defined, which may also be gradient) is close to 0. That gradient is a measure of change, which is a complementary of stability. 
- Still need to think how this is related to strictly local gradient, which defines our blobs. The points with very low 2nd gradient (Laplacian) are similar to extrema, but not exclusive. That was incremental-derivation fork in old 3-fork version of intra_blob: https://github.com/boris-kz/CogAlg/blob/master/frame_2D_alg/Illustrations/intra_blob_3_fork_scheme.png 
Or incremental-range fork, it is selective for very low gradient.

- Orientation assignment in SIFT is by angle of gradient, same as ours? Could you look at corner detectors, it sounds like our comp_angle? Corner would be a local minimum of ma (match of angle), although we currently compute deviation, not extrema.

**Reply by kwcckw** [reply]
okay, let me go through all of these first and i will check on orientation and those corners again.

**Reply by kwcckw** [reply]
> 
> 
>     * I don't want to use Gaussians because it's a form of resolution reduction. I think resolution (a type of my filters) should only be reduced after evaluation per pattern, which means after a full cross-comp and clustering cycle. So, scale-space here is compositional hierarchy of patterns, not a spectrum of Gaussian kernels.
> 
>     * Key points in SIFT are just another name for extrema? They may be more stable than blob averages / medians because their gradient (of the parameter over which the extrema is defined, which may also be gradient) is close to 0. That gradient is a measure of change, which is a complementary of stability.

Right, but they are computed via difference of gaussian (DoG), so it's the difference of 2 gaussian processed image, with different gaussian kernel value. I guess this is not suitable in our use case since the difference is obtained via the reduced image, so some information might be lost.

> 
>     * Still need to think how this is related to strictly local gradient, which defines our blobs. The points with very low 2nd gradient (Laplacian) are similar to extrema, but not exclusive. That was incremental-derivation fork in old 3-fork version of intra_blob: https://github.com/boris-kz/CogAlg/blob/master/frame_2D_alg/Illustrations/intra_blob_3_fork_scheme.png
>       Or incremental-range fork, it is selective for very low gradient.
> 
>     * Orientation assignment in SIFT is by angle of gradient, same as ours?

It is computed via gradient direction but the final value is approximated into 36 bins of histogram, so may not be the same in our case.
Ref: https://medium.com/data-breach/introduction-to-sift-scale-invariant-feature-transform-65d7f3a72d40
![image](https://user-images.githubusercontent.com/52521979/106355420-b15b5700-6332-11eb-924e-5a4d4d08bf0d.png)


> Could you look at corner detectors, it sounds like our comp_angle? Corner would be a local minimum of ma (match of angle), although we currently compute deviation, not extrema.

From my initial checking on the pseudo code, looks like it's similar but different at the point when they are using extrema while we are using the sum of difference . First they use sobel filter to get x and y gradient, then they find the local maxima as corner by using 3x3 window filter. So this would be different with the sum of deviation in our case.
Ref: https://medium.com/data-breach/introduction-to-harris-corner-detector-32a88850b3f6

**Reply by boris-kz** [reply]
So, low deviation of gradient (-G) is not sufficient as a measure of extrema, it may also indicate flat areas.
But we also compute SAD as an inverse measure of match. 

High deviation of SAD (-M) corresponds to either steep slope or steep extrema, and the later is when it is combined with low G. Thus, specific measure of extrema = -G * -M: directionally-compensated variation?

This is not necessarily exclusive max | min by design, but I think it will be effectively pretty exclusive anyway?
Still not sure how to relate this measure of extrema to separate G and M we are using now.  
(I wrote it before seeing Chee's post).

**Reply by boris-kz** [reply]
> 
> From my initial checking on the pseudo code, looks like it's similar but different at the point when they are using extrema while we are using the sum of difference . First they use sobel filter to get x and y gradient, then they find the local maxima as corner by using 3x3 window filter. So this would be different with the sum of deviation in our case.
> Ref: https://medium.com/data-breach/introduction-to-harris-corner-detector-32a88850b3f6

Thanks Chee! So, I think my description above is a better way to quantify extrema, and they don't have to be singular.

**Reply by kwcckw** [reply]
> 
> 
> > From my initial checking on the pseudo code, looks like it's similar but different at the point when they are using extrema while we are using the sum of difference . First they use sobel filter to get x and y gradient, then they find the local maxima as corner by using 3x3 window filter. So this would be different with the sum of deviation in our case.
> > Ref: https://medium.com/data-breach/introduction-to-harris-corner-detector-32a88850b3f6
> 
> Thanks Chee! So, I think my description above is a better way to quantify extrema, and they don't have to be singular.

Right, but i think those extrema also can start with some singular point? Then by grouping them with some methodology, they could be sorted into certain form of extrema?

**Reply by kwcckw** [reply]
> 
> 
> So, low deviation of gradient (-G) is not sufficient as a measure of extrema, it may also indicate flat areas.
> But we also compute SAD as an inverse measure of match.
> 
> High deviation of SAD (-M) corresponds to either steep slope or steep extrema, and the later is when it is combined with low G. Thus, specific measure of extrema = -G * -M: directionally-compensated variation?
> 
> This is not necessarily exclusive max | min by design, but I think it will be effectively pretty exclusive anyway?
> Still not sure how to relate this measure of extrema to separate G and M we are using now.
> (I wrote it before seeing Chee's post).

At this point i'm also don't have a better suggestion yet, and i cannot sure whether -G * -M is suitable or not. However, let say we want to test it out, how do you intend to compare and evaluate which method is more suitable? By output some visualization with image data? Or just test with some simple data array?

**Comment by boris-kz** [comment]
Singular extrema make things a lot simpler, but I think this simplification is artificial. Replacing them with blobs of above-average values is a lot more accurate, I can think of no good reason for non-maximum suppression? Those values should represent approximate peaks: areas of directionally-symmetrical (low G) high local variation (SAD). 

So, it should be (ave_G - G) * SAD - ave_product_of_the_two? 
These areas maybe more stable than simply low-SAD (high M) areas because it's a kind of internal contrast: 
G is low despite SAD being high. Still haven't quite conceptualized it yet.

Yes, it's best to visualize these blobs with real images.

**Reply by boris-kz** [reply]
> These areas maybe more stable than simply low-SAD (high M) areas because it's a kind of internal contrast:
> G is low despite SAD being high. Still haven't quite conceptualized it yet.

Basically, high SAD & low G means that local variation is directionally reversed. If such reversal is consistent (resulting blob is large), we can assume that there is some reversing / stabilizing force: ambient pressure or gravity. This stabilizing force is a meta-pattern, it enforces locally low G, so it's more valuable resulting low-G patterns.  Still speculative.

**Reply by kwcckw** [reply]
> 
> 
> Singular extrema make things a lot simpler, but I think this simplification is artificial. Replacing them with blobs of above-average values is a lot more accurate, I can think of no good reason for non-maximum suppression? 

Yes singular extrema is simple and straightforward. Typically you would get a group of extrema  points in certain area, and by using NMS (non-maximum supression), you are able to remove those redundant points or  'noise' . The result would be cleaner but we may discard some of the important points as well.  In our case, an area of extrema points would be more relevant.

> Those values should represent approximate peaks: areas of directionally-symmetrical (low G) high local variation (SAD).
> 
> So, it should be (ave_G - G) * SAD - ave_product_of_the_two?
> These areas maybe more stable than simply low-SAD (high M) areas because it's a kind of internal contrast:
> G is low despite SAD being high. Still haven't quite conceptualized it yet.
> 
> Yes, it's best to visualize these blobs with real images.

Maybe we can test with images and evaluate again based on the visualized result?

**Reply by boris-kz** [reply]
Yes, it would be interesting to compare with G blobs. G has a contrast value: presumed / computed borrow from adjacent flat blobs. Local extrema have G direction-reversal, basically directional contrast. In terms of predictive value, they will be borrowing from presumed surrounding high-G (direction-consistent) blobs. 
   
So, we can add e-fork (extrema or reversal) in intra-blob. 
That will be called for -G_blobs if (M/A + G(negative) /A) > ave_e. No idea where to set ave_e, 
It will be cluster_sub_eval -> flood_fill only, no intra_comp calls. 

Actually, it will have to go through comp_r first, this is meaningless for 2x2 kernels.
Hold off on it for now.

**Reply by kwcckw** [reply]
> 
> 
> Yes, it would be interesting to compare with G blobs. G has a contrast value: presumed / computed borrow from adjacent flat blobs. Local extrema have G direction-reversal, basically directional contrast. In terms of predictive value, they will be borrowing from presumed surrounding high-G (direction-consistent) blobs.
> 
> So, we can add e-fork (extrema or reversal) in intra-blob.
> That will be called for -G_blobs if (M/A + G(negative) /A) > ave_e. No idea where to set ave_e,
> It will be cluster_sub_eval -> flood_fill only, no intra_comp calls.
> 
> Actually, it will have to go through comp_r first, this is meaningless for 2x2 kernels.
> Hold off on it for now.

Alright, but it would be quite difficult to precisely visualize the sub blobs in intra_blob. The simplest way is to just apply the whole image to intra_comp or entire e- fork, similar with what we did previously, this should be the quickest way if you want to visualize them without much effort.

**Reply by boris-kz** [reply]
Ok, use draw_intra_comp version that only calls comp_r(5x5), 
and then draws values of e_ = m_- 1.41*g_?
Both m and g are absolute here, no - ave.

**Reply by kwcckw** [reply]
> 
> 
> Ok, use draw_intra_comp version that only calls comp_r(5x5),
> and then draws values of e_ = m_- 1.41*g_?
> Both m and g are absolute here, no - ave.

okay, but for comp_r, we may need to use interpolation a bit since the output size would be different with input size. I will look at this and test again some time later.

**Reply by kwcckw** [reply]
from my initial testing, here are the outputs:

![image](https://user-images.githubusercontent.com/52521979/107489628-cef6af00-6bc3-11eb-906c-2db0b1fcb84f.png)

for e_, why would we need the *1.41 on g ( e_ = m_- 1.41*g_) ? Without the 1.41, it would looks better:

with *1.41:
e_ = m_- 1.41*g_

![image](https://user-images.githubusercontent.com/52521979/107489773-fd748a00-6bc3-11eb-9d64-b46529a2bef4.png)


without *1.41:
e_ = m_- *g_

![image](https://user-images.githubusercontent.com/52521979/107489826-0e250000-6bc4-11eb-9c70-cc00dc22ecb1.png)

**Reply by boris-kz** [reply]
Thanks a lot Chee! I think 1.41 is the ratio of SAD to G: hypot (Gy, Gx), so * 1.41 is supposed to make m and g comparable.
We may have to try it with different kernel sizes, not sure how to chose them.

**Reply by kwcckw** [reply]
> 
> 
> Thanks a lot Chee! I think 1.41 is the ratio of SAD to G: hypot (Gy, Gx), so * 1.41 is supposed to make m and g comparable.

So hypot(Gy,Gx) = sqrt(  Gy^2 + Gx^2  ）

and

1.41 = sqrt(  1^2 + 1^2  ）

I don't see how 1.41 is related to hypot(Gy,Gx)? 

> We may have to try it with different kernel sizes, not sure how to chose them.

Maybe i will perform some experiments based on different kernel size and ratio of SAD, but for the evaluation method, i think we can just evaluate based on the output images now.

**Reply by boris-kz** [reply]
It doesn't hold in every case, but I think average SAD = 1.41 average G. 
How would you define this ratio? It's basic math, we shouldn't have to experiment.

**Reply by kwcckw** [reply]
> 
> 
> It doesn't hold in every case, but I think average SAD = 1.41 average G.
> How would you define this ratio?

So for G, it is computed from 2 directional elements and it is computed via square root of square sum.

for SAD, we are computing it from the sum of 8 directional differences. 

So maybe 1.41 /4? Since SAD is using 8 directions while G is using 2 directions?

> It's basic math, we shouldn't have to experiment.

Maybe just experiment with different kernel size and see the trend of the output images?

**Reply by boris-kz** [reply]
Oh, ok I was thinking of 2x2 kernel, where there are only 2 directions.
So, in 3x3 it's 1.41 / 4, for the case where Gy and Gx are equal.
But to the extent that they are not equal, I think the ratio is lower than 1.41?
BTW, I think it should look better with 5x5 kernels.

**Reply by kwcckw** [reply]
> 
> 
> Oh, ok I was thinking of 2x2 kernel, where there are only 2 directions.
> So, in 3x3 it's 1.41 / 4, for the case where Gy and Gx are equal.
> But to the extent that they are not equal, I think the ratio is lower than 1.41?
> BTW, I think it should look better with 5x5 kernels.

right, the ratio should be lower than 1.41. Also there are some mistakes in the previous code, when rng = 2, there should be 16 directional images and the directional images should be increases when the rng increases. I will post here when i have the updated results.

**Reply by kwcckw** [reply]
from my testing, i can see the edges in e_ become more more distinct when the rng increases, but the trade-off is losing information since we lost more information when the rng increases.

rng=1
![image](https://user-images.githubusercontent.com/52521979/107763021-dd320000-6d68-11eb-9d56-4070ed6ae67b.png)

rng=2
![image](https://user-images.githubusercontent.com/52521979/107763073-f470ed80-6d68-11eb-8ac5-d19bb29a55b2.png)

rng=3
![image](https://user-images.githubusercontent.com/52521979/107763168-208c6e80-6d69-11eb-8fd0-4323ec950a2a.png)

rng=4
![image](https://user-images.githubusercontent.com/52521979/107763185-271ae600-6d69-11eb-809c-c765b9c2e4bc.png)

rng=5
![image](https://user-images.githubusercontent.com/52521979/107763198-2d10c700-6d69-11eb-85ff-e180b3ba9712.png)

rng=6
![image](https://user-images.githubusercontent.com/52521979/107763205-31d57b00-6d69-11eb-95ca-26786b8bb9fe.png)

rng=7
![image](https://user-images.githubusercontent.com/52521979/107763214-3732c580-6d69-11eb-8c12-d6b711611789.png)

rng=8
![image](https://user-images.githubusercontent.com/52521979/107763226-3b5ee300-6d69-11eb-85b0-53f7fe95fb81.png)

rng=9
![image](https://user-images.githubusercontent.com/52521979/107763241-41ed5a80-6d69-11eb-93fb-a8f68365ff07.png)

rng=10
![image](https://user-images.githubusercontent.com/52521979/107763250-46b20e80-6d69-11eb-87ac-f521387b05dd.png)

**Comment by boris-kz** [comment]
Thanks Chee! I think resolution is lost by design, your shift_img does the same skipping we built in comp_r? 
That made sense when comp_r was designed for flat blobs, but now it's performed to analyze high-variation (high-m) areas.
So, we need comp_rng without skipping: it should compute kernels for each pixel, without output interpolation?

**Reply by kwcckw** [reply]
> 
> 
> Thanks Chee! I think resolution is lost by design, your shift_img does the same skipping we built in comp_r?

Yes, it shifts and skip based on the rng number, it should be similar with our current work.

If rng = 1 (3x3), distance between kernel = 2, number of shifted images = 8
If rng = 2 (5x5), distance between kernel = 3, number of shifted images = 16
If rng = 3 (7x7), distance between kernel = 4, number of shifted images = 24
and so on.

> That made sense when comp_r was designed for flat blobs, but now it's performed to analyze high-variation (high-m) areas.
> So, we need comp_rng without skipping: it should compute kernels for each pixel, without output interpolation?

With this, the resolution should be better, i will try that later. Also for the current code, i'm using diametrically opposed images to compute the difference instead of using differences between center image and each directional image, do you think differences between center image and each directional image will be better?

**Reply by boris-kz** [reply]
m should definitely be computed for each direction separately, g is probably fine with diametrically opposed comparison.

**Reply by kwcckw** [reply]
okay, right now dy__, dx__ and g are computed via diametrically opposed comparison, while m is computed via each directional differences. So i guess no changes are needed on those computation.

**Reply by kwcckw** [reply]
without the skipping, the output when rng increases would looks like those sobel kernel with larger kernel size:

rng = 1
![image](https://user-images.githubusercontent.com/52521979/107846438-c0580400-6e1e-11eb-84c3-dce4f5c3cb61.png)

rng = 2
![image](https://user-images.githubusercontent.com/52521979/107846443-c8b03f00-6e1e-11eb-92a3-e7e481842923.png)

rng = 3
![image](https://user-images.githubusercontent.com/52521979/107846447-cf3eb680-6e1e-11eb-9f69-8ee0815abbbc.png)

rng = 4
![image](https://user-images.githubusercontent.com/52521979/107846450-d5cd2e00-6e1e-11eb-9567-989c953c7f65.png)

rng = 5
![image](https://user-images.githubusercontent.com/52521979/107846457-dcf43c00-6e1e-11eb-8f1c-68cae0035210.png)

rng = 6
![image](https://user-images.githubusercontent.com/52521979/107846459-e5e50d80-6e1e-11eb-97e6-7d0d20aadfbd.png)

rng = 7
![image](https://user-images.githubusercontent.com/52521979/107846462-ebdaee80-6e1e-11eb-8761-de7987f7c286.png)

rng = 8
![image](https://user-images.githubusercontent.com/52521979/107846464-f1383900-6e1e-11eb-94a2-10bf5e0f6f42.png)

rng = 9
![image](https://user-images.githubusercontent.com/52521979/107846465-f72e1a00-6e1e-11eb-8193-bb328e563cde.png)

rng = 10
![image](https://user-images.githubusercontent.com/52521979/107846467-fd23fb00-6e1e-11eb-82b6-5bd71df918d9.png)

**Comment by kwcckw** [comment]
gr also looks quite interesting, but edges in e_ looks cleaner:

![image](https://user-images.githubusercontent.com/52521979/107846543-85a29b80-6e1f-11eb-844f-a48cbad5e636.png)

![image](https://user-images.githubusercontent.com/52521979/107846585-d6b28f80-6e1f-11eb-8233-1ee3068c317b.png)

**Reply by boris-kz** [reply]
Thanks, gr *is* Sobel. But it's hard to compare with e_ because highest intensity g is black, while e is white, could you make black too?

**Reply by kwcckw** [reply]
Here are the images. Looking at it, e__ is very similar to m__, but there are some values difference if we subtract them with each other. Also it's very obvious the edges in e__ is better than g__.

[intra_comp0.zip](https://github.com/boris-kz/CogAlg/files/5975784/intra_comp0.zip)

**Reply by boris-kz** [reply]
So, in this image we have unidirectional edges, which means that  m is is generally proportional to g. That's why there is very little difference between m and e. We can make it bigger by using g*2 or *3? 
It should be different with bidirectional edges (ridges): thin dark / light lines on light / dark background. Maybe hair strands in raccoon, or text, using large kernels?

But m (SAD) does look a lot better than g for edge detection, do you know why people use g?
BTW, it would look better with dark contours: displayed intensity should be inverse to g / m / e.

**Reply by kwcckw** [reply]
> 
> 
> So, in this image we have unidirectional edges, which means that m is is generally proportional to g. That's why there is very little difference between m and e. We can make it bigger by using g*2 or *3?

actually i tried with raccoon image and using g*3 in computing e__, i can see edges in e__  looks better a bit comparing with m__ now, but i'm not sure whether is the effect of scaling since i'm scaling all negative values to positive. 
[intra_comp0.zip](https://github.com/boris-kz/CogAlg/files/5977900/intra_comp0.zip)


> It should be different with bidirectional edges (ridges): thin dark / light lines on light / dark background. Maybe hair strands in raccoon, or text, using large kernels?

Yes, i'm testing with rng of 1 to 10, and with input raccoon image. Now, e__ looks better when rng is low (rng = 1), but when rng gets bigger, it's hardly to see any difference between e__ and m__.


> 
> But m (SAD) does look a lot better than g for edge detection, do you know why people use g?

I don't have much info on this, and actually i never heard of SAD until now. Maybe it's called another term but i'm didn't aware of it. Or it's just don't have any people using this approach in any of the published work yet.

> BTW, it would look better with dark contours: displayed intensity should be inverse to g / m / e.

Yes, i just pushed another edit to display  dark contours and bright edges.

**Reply by boris-kz** [reply]
So, m - g is distinct when g cancels-out within a kernel: we have two edges with opposite angle, or thin ridge.
But that only works if kernel size ~ ridge width. Most images have ridges of various width, so the effect is hardly noticeable for any given kernel size. The exception would be text: letters are composed of fixed-width lines. 

>  Maybe it's called another term but i'm didn't aware of it. Or it's just don't have any people using this approach in any of the published work yet. 

That's highly unlikely, SAD is an established term, and simpler to compute than G. G is useful when you need to consider direction, but we can still use SAD for magnitude.

**Reply by kwcckw** [reply]
Yes you are right. I found some previous works on SAD and most of them related to matching application.

https://www.researchgate.net/publication/220501047_Faster_image_template_matching_in_the_sum_of_the_absolute_value_of_differences_measure 

https://www.sciencedirect.com/science/article/pii/S2215016120303149

However,  a brief reading in wiki shows that SAD can be used for object recognition as well, but need some edge detection to complement it:
![image](https://user-images.githubusercontent.com/52521979/108160111-6196d100-7123-11eb-8841-dce7634068a6.png)
https://en.wikipedia.org/wiki/Sum_of_absolute_differences

> So, m - g is distinct when g cancels-out within a kernel: we have two edges with opposite angle, or thin ridge.
> But that only works if kernel size ~ ridge width. Most images have ridges of various width, so the effect is hardly noticeable > for any given kernel size. The exception would be text: letters are composed of fixed-width lines

so if the kernel size is too small and ridge size is large, it might not able to detect the edges since there is no distinct changes detected. And if the kernel size is large and ridge size is small, the detected edges may not notice-able or distinct enough. So maybe a feedback system to change the kernel size dynamically after we perform some 1st level of processing on certain section of image (blob by blob)?

**Reply by boris-kz** [reply]
That paper is on template matching, not edge detection.

> So maybe a feedback system to change the kernel size dynamically after we perform some 1st level of processing on certain section of image (blob by blob)?

Ridge width also changes within a blob. So, I don't see any consistent way to use e, except for text and hand drawings. 
It may not be needed, we will combine the two edges of a ridge after comparing them, in comp_blob.

**Reply by kwcckw** [reply]
> 
> 
> That paper is on template matching, not edge detection.
> 
> > So maybe a feedback system to change the kernel size dynamically after we perform some 1st level of processing on certain section of image (blob by blob)?
> 
> Ridge width also changes within a blob. So, I don't see any consistent way to use e, except for text and hand drawings.
> It may not be needed, we will combine the two edges of a ridge after comparing them, in comp_blob.

Alright, maybe after we have some initial results in comp_blob, we can compare their output and evaluate again.
